api/v1alpha1/statuscheck_types.go

```diff
@@ -0,0 +1,215 @@
+	// +optional
+	CompletionTime *metav1.Time `json:"completionTime,omitempty"`
+
+	// Count represents the total number of the status check executed.
+	// +optional
+	Count int64 `json:"count,omitempty"`
+
+	// Conditions represents the latest available observations of a StatusCheck's current state.
+	// +optional
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	Conditions []StatusCheckCondition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type"`
+
+	// Records contains the history of the execution of StatusCheck.
+	// +optional
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	Records []StatusCheckRecord `json:"records,omitempty" patchStrategy:"merge" patchMergeKey:"type"`
+}
+
+type StatusCheckOutcome string
+
+const (
+	StatusCheckOutcomeSuccess StatusCheckOutcome = "Success"
+	StatusCheckOutcomeFailure StatusCheckOutcome = "Failure"
+)
+
+type StatusCheckRecord struct {
+	ProbeTime *metav1.Time       `json:"probeTime"`
+	Outcome   StatusCheckOutcome `json:"outcome"`
+}
+
+type StatusCheckConditionType string
+
+const (
+	// StatusCheckConditionComplete means the status check has completed its execution.
+	StatusCheckConditionComplete StatusCheckConditionType = "Complete"
+	// StatusCheckConditionFailed means the status check has failed its execution.
+	StatusCheckConditionFailed StatusCheckConditionType = "Failed"
+)
+
+type StatusCheckReason string
+
+const (
+	StatusCheckSuccess StatusCheckReason = "StatusCheckSuccess"
+	// TODO add more reason when implementing StatusCheck
+)
+
+type StatusCheckCondition struct {
+	Type               StatusCheckConditionType `json:"type"`
+	Status             corev1.ConditionStatus   `json:"status"`
+	Reason             StatusCheckReason        `json:"reason"`
+	LastProbeTime      *metav1.Time             `json:"lastProbeTime"`
+	LastTransitionTime *metav1.Time             `json:"lastTransitionTime"`
+}
+
+type EmbedStatusCheck struct {
+	// +optional
+	HTTPStatusCheck *HTTPStatusCheck `json:"http,omitempty"`
+}
+
+type HTTPHeaderPair struct {
+	Name  string `json:"name"`
+	Value string `json:"value"`
+}
+
+type HTTPCriteria struct {
+	// StatusCode defines the expected http status code for the request.
+	// A statusCode string could be a single code (e.g. 200),
+	// or a range (e.g. 200-400).
+	StatusCode string `json:"statusCode,omitempty" webhook:"StatusCode"`
+	// TODO: support response body
+}
+
+type HTTPRequestMethod string
+
+const (
+	MethodGet  = "GET"
+	MethodPost = "POST"
+)
+
+type HTTPStatusCheck struct {
+	RequestUrl string `json:"url"`
+
+	// +optional
+	// +kubebuilder:validation:Enum=GET;POST
+	// +kubebuilder:default=GET
+	RequestMethod HTTPRequestMethod `json:"method,omitempty"`
+	// +optional
+	RequestHeaders []HTTPHeaderPair `json:"headers,omitempty"`
+	// +optional
+	RequestBody string `json:"body,omitempty"`
+	// Criteria defines how to determine the result of the status check.
+	Criteria HTTPCriteria `json:"criteria,omitempty"`
```

Comment:
we do not need to add `omitempty` on the required fields

Reply:
Updated.